nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutSQL.java

```diff
@@ -134,6 +134,14 @@
             .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
             .build();
 
+    static final PropertyDescriptor AUTO_COMMIT = new PropertyDescriptor.Builder()
+            .name("database-session-autocommit")
+            .displayName("Database session autocommit value")
+            .description("The autocommit mode to set on the database connection being used.")
+            .allowableValues("true", "false")
+            .defaultValue("false")
+            .build();
```

Comment:
@viswaug IIRC disable auto-commit is required to support "Support Fragmented Transactions" and/or "Rollback On Failure".  We need to implement `customValidate` method, too, if we're going to make auto-commit configurable. If auto-commit is enabled, the processor should not allow using "Support Fragmented Transactions" and/or "Rollback On Failure".

Reply:
@ijokarumawak gotcha. how about i remove the property i just added.  and set autocommit to true when the SUPPORT_TRANSACTIONS property value is set to false?

Reply:
@viswaug That approach may work, too. However, I prefer exposing auto-commit property and let user to enable it explicitly. And protect existing logics work as expected by adding following custom validation:
- if auto-commit is enabled:
    - `Support Fragmented Transactions` should be false
    - && `Rollback On Failure` should be false

This way even if we forgot about other conditions that require auto-commit, user can disable auto-commit to work-around. PutSQL has been there for long time and used by so many flows. We can not introduce any degrading issue by supporting auto-commit. 


BTW, the description of Snowflake database issue is not clear enough to me on how auto-commit setting relates to the issue.

> This is causing an issue with the snowflake DB where abruptly disconnected sessions do not release the locks they have taken.

Do you have any existing Snowflake issue or resource that we can refer? Or if not have you consider reporting the issue to Snowflake project?

And also, if you have any use-case (other than avoiding Snowflake issue) where auto-commit is preferable, please add such to auto-commit property description.

Reply:
@ijokarumawak  Here is the snowflake documentation that refers to the locks being held for ever during abruptly disconnected sessions. 
https://docs.snowflake.net/manuals/sql-reference/transactions.html#aborting-transactions

I have also confirmed this with the snowflake support ticket and the resolution suggested was to set the autocommit value to true. 

> I did reviewed the query id thanks for providing it. The query is associated to session id 2392474452590 and from the session id we found that there was an alter session set autocommit=false was executed (query id for your reference 10d44ff5-69a4-4d8c-91c3-a206c4a126b8) and after that there was no commit executed explicitly. This lead to the open transactions in the session and hence there was a lock created. 

> By default autocommit is true and, once all the transaction is completed, it gets automatically committed. The commit query will not be visible in the history as this is a background task. For best practices, we recommend to manually commit the transactions if the session has set to autocommit=false . 

> Yes, there is a difference between a session being "closed" versus "terminated abruptly". Session being "closed" implies to those sessions which is closed manually after all the transactions is completed . Session being "terminated abruptly" implies to those sessions which terminates due to various reason like network issues, system outage ..etc. 

I am done making the changes you had requested. I will have a PR out soon. I just need to hone my git skills to combine these commits and send a PR ... still figuring that part out ...

Reply:
@viswaug Thanks for the snowflake doc link and explanation. Look forward to see new commits for further review. In order to update your PR, you just need to:
1. Make code changes
2. Add the changed files for a commit: `git add <file>` or `git add -A` (all changed files)
3. Commit with some comments: `git commit`, this makes a commit to your local branch
4. Push the commit to your remote branch: `git push origin configurable_autocommit_putsql`. This command adds the new commit to this PR.